internal/luks2/metadata.go

```diff
@@ -388,64 +397,96 @@ func (c *Config) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
+type rawToken struct {
+	typ  TokenType
+	data []byte
+}
+
+func (t *rawToken) UnmarshalJSON(data []byte) error {
+	var d struct {
+		Type TokenType
+	}
+	if err := json.Unmarshal(data, &d); err != nil {
+		return err
+	}
+
+	t.typ = d.Type
+	t.data = data
+	return nil
+}
+
 // Token corresponds to a token object in the JSON metadata of a LUKS2 volume. It
-// describes how to retrieve a passphrase or key for a keyslot.
-type Token struct {
-	Type     string                 // Token type ("luks2-" prefixed types are reserved for cryptsetup)
+// describes how to retrieve a passphrase or key for a keyslot. Tokens decoded by
+// ReadHeader will be represented by a type-specific implementation if a
+// TokenHandler is registered for it, or GenericToken.
+type Token interface {
+	GetType() TokenType // Token type ("luks2-" prefixed types are reserved for cryptsetup)
+	GetKeyslots() []int // Keyslots assigned to this token
+}
+
+// TokenHandler provides a mechanism for an external package to decode
+// custom token types.
+type TokenHandler func([]byte) (Token, error)
+
+var tokenHandlers = make(map[TokenType]TokenHandler)
+
+// GenericToken corresponds to a token that doesn't have a more type-specific
+// representation.
+type GenericToken struct {
+	Type     TokenType              // Token type ("luks2-" prefixed types are reserved for cryptsetup)
```

Comment:
this could be TokenType and KeyslotIndexes for example or something like that

Reply:
I've renamed these now to TokenType and TokenKeyslots.